src/components/i18n/FormattedCompMessage.js

```diff
@@ -0,0 +1,197 @@
+ * its translations from react intl as well. The FormattedCompMessage component can
+ * be used wherever it is valid to put JSX text. In regular Javascript
+ * code, you should continue to use the intl.formatMessage() call and
+ * extract your strings into a message.js file.
+ */
+class FormattedCompMessage extends React.Component<Props, State> {
+    composition: Composition;
+
+    str: string;
+
+    static defaultProps = {
+        tagName: 'span',
+    };
+
+    constructor(props: Props) {
+        super(props);
+
+        // these parameters echo the ones in react-intl's FormattedMessage
+        // component, plus a few extra
+        const {
+            defaultMessage, // The English string + HTML + components that you want translated
+            count, // the pivot count to choose a plural form
+            children, // the components within the body
+        } = this.props;
+
+        const sourceElements = defaultMessage || children;
+
+        if (sourceElements) {
+            const composition = new Composition(sourceElements);
+            let source = '';
+
+            if (!isNaN(Number(count))) {
+                if (children) {
+                    source = this.composePluralString(children);
+                } else if (isDevEnvironment()) {
+                    throw new Error('Cannot use count prop on a FormattedCompMessage component that has no children.');
+                }
+            } else {
+                source = composition.compose();
+            }
+
+            this.state = {
+                source,
+                composition,
+            };
+        }
+    }
+
+    /**
+     * Search for any Plural elements in the children, and
+     * then construct the English source string in the correct
+     * format for react-intl to use for pluralization
+     * @param {React.Element} children the children of this node
+     * @return {string} the composed plural string
+     */
+    composePluralString(children) {
+        const categories = {};
+        React.Children.forEach(children, child => {
+            if (typeof child === 'object' && React.isValidElement(child) && child.type.name === 'Plural') {
+                const childComposition = new Composition(child.props.children);
+                categories[child.props.category] = childComposition.compose();
+            }
+        });
+        if (!categories.one || !categories.other) {
+            if (isDevEnvironment()) {
+                throw new Error(
+                    'Cannot use count prop on a FormattedCompMessage component without giving both a "one" and "other" Plural component in the children.',
+                );
+            }
+        }
+        // add these to the string in a particular order so that
+        // we always end up with the same string regardless of
+        // the order that the Plural elements were specified in
+        // the source code
+        const categoriesString = [
```

Comment:
Can this live as its own constant in this file rather than being re-instantiated each time?